src/v/cluster/tm_stm_types.h

```diff
@@ -92,7 +182,186 @@ struct tx_metadata {
     std::chrono::milliseconds get_timeout() const { return timeout_ms; }
 
     bool delete_partition(const tx_partition& part);
+    /**
+     * Validates and updates the transaction metadata status if transition is
+     * allowed by transaction FSM.
+     */
+    std::optional<state_transition_error> try_update_status(tx_status);
+    /**
+     * Returns true if transaction is already finished. When transaction is
+     * finished it is either commit or abort completed and has no participants
+     */
+    bool is_finished() const;
 
     friend std::ostream& operator<<(std::ostream&, const tx_metadata&);
 };
+
+bool is_state_transition_valid(const tx_metadata&, tx_status);
+
+/**
+ * Deprecated metadata formats
+ */
+
+// Updates in v1.
+//  + last_update_ts - tracks last updated ts for transactions expiration
+//  + tx_status::tombstone - Removes all txn related state upon apply.
+struct transaction_metadata_v1 {
+    static constexpr uint8_t version = 1;
+
+    enum tx_status : int32_t {
+        ongoing,
+        preparing,
+        prepared,
+        aborting, // abort is initiated by a client
+        killed,   // abort is initiated by a timeout
+        ready,
+        tombstone,
+    };
+
+    struct tx_partition {
+        model::ntp ntp;
+        model::term_id etag;
+
+        bool operator==(const tx_partition& other) const = default;
+    };
+
+    struct tx_group {
+        kafka::group_id group_id;
+        model::term_id etag;
+    };
+
+    // id of an application executing a transaction. in the early
+    // drafts of Kafka protocol transactional_id used to be named
+    // application_id
+    kafka::transactional_id id;
+    // another misnomer in fact producer_identity identifies a
+    // session of transactional_id'ed application as any given
+    // moment there maybe only one session per tx.id
+    model::producer_identity pid;
+    // tx_seq identifues a transactions within a session so a
+    // triple (transactional_id, producer_identity, tx_seq) uniquely
+    // identidies a transaction
+    model::tx_seq tx_seq;
+    // term of a transaction coordinated started a transaction.
+    // transactions can't span cross term to prevent loss of information stored
+    // only in memory (partitions and groups).
+    model::term_id etag;
+    tx_status status;
+    std::chrono::milliseconds timeout_ms;
+    ss::lowres_system_clock::time_point last_update_ts;
+    std::vector<tx_partition> partitions;
+    std::vector<tx_group> groups;
+
+    cluster::tx_status upcast(tx_status status) {
+        switch (status) {
+        case tx_status::ongoing:
+            return cluster::tx_status::ongoing;
+        case tx_status::preparing:
+            return cluster::tx_status::preparing_commit;
+        case tx_status::prepared:
+            return cluster::tx_status::completed_commit;
```

Comment:
I had an impression that `prepared` is actually equivalent to `preparing_commit` and `preparing` is unused (?)